File: Xenon/Xenon/Upgrade.ts
import gsCVector = require("Vector");
import gsCMap = require("Map");
import CActor = require("Actor");
import CScene = require("Scene");
import CWeapon = require("Weapon");
import CMissileWeapon = require("MissileWeapon");
import CHomingMissileWeapon = require("HomingMissileWeapon");
import CLaserWeapon = require("LaserWeapon");
import enums = require("Enums");
import CShip = require("Ship");
import Pickup = require("Pickup");
import CPlayGameState = require("PlayGameState");

class CUpgrade extends CActor {

    //-------------------------------------------------------------
    UPGRADE_MAP_HIT: number = 10;			// energy lost if upgrade hits map
    //-------------------------------------------------------------

    m_offset: gsCVector;
    m_weapon: CWeapon;
    m_weapon_type: enums.WeaponType;
    protected m_playGameState: CPlayGameState;

    constructor(scene?: CScene) {
        super(scene);
        this.m_offset = new gsCVector(0.0, 0.0);
        this.m_weapon = null;
        this.m_weapon_type = enums.WeaponType.NO_WEAPON;
        this.m_name = "Upgrade";
    }

    //-------------------------------------------------------------

    public activate() {
        if (!this.isActive()) {
            this.setWeapon(enums.WeaponType.MISSILE_WEAPON, 0);
        }
        return super.activate();
    }

    //-------------------------------------------------------------

    public kill() {
        if (this.m_weapon) {
            this.m_weapon.kill();
            this.m_weapon = null;
        }
        super.kill();
    }

    //-------------------------------------------------------------

    public setOffset(offset: gsCVector) {
        this.m_offset = offset;
    }

    //-------------------------------------------------------------

    public registerHit(energy: number, hitter: CActor) {
        if (this.getOwner() && (<CShip>this.getOwner()).getDiveLevel() > 0) {
            return;
        }
        //this.registerHit(energy, hitter);
        super.registerHit(energy, hitter);
    }

    //-------------------------------------------------------------

    public onCollisionWithActor(actor: CActor) {
        if (this.getOwner() &&
            (<CShip>this.getOwner()).getDiveLevel() > 0) {
            return;
        }

        switch (actor.getActorInfo().m_type) {
            case enums.ActorType.ACTOR_TYPE_PICKUP:
                var act = <Pickup.CPickup>actor; //(<Pickup.CPickup> actor).collect();
                act.collect();
                actor.kill();
                break;
            case enums.ActorType.ACTOR_TYPE_ALIEN:
                this.registerHit(1, this);
                actor.registerHit(1, this);
                break;
        }
    }

    //-------------------------------------------------------------

    public onCollisionWithMap(map: gsCMap, hits: number) {
        if (this.getOwner() &&
            (<CShip>this.getOwner()).getDiveLevel() > 0) {
            return;
        }
        this.registerHit(this.UPGRADE_MAP_HIT, this);
    }

    //-------------------------------------------------------------

    //public setWeapon(type: enums.WeaponType, grade: enums.WeaponGrade = enums.WeaponGrade.WEAPON_STANDARD) {
    public setWeapon(type: enums.WeaponType, grade?: enums.WeaponGrade) {
        if (this.m_weapon) {
            this.m_weapon.kill();
            this.m_weapon = null;
        }

        this.m_weapon_type = type;

        switch (this.m_weapon_type) {
            case enums.WeaponType.NO_WEAPON:
                this.m_weapon = null;
                break;
            case enums.WeaponType.MISSILE_WEAPON:
                this.m_weapon = new CMissileWeapon(this.m_scene, this.m_playGameState);
                break;
            case enums.WeaponType.HOMING_MISSILE_WEAPON:
                this.m_weapon = new CHomingMissileWeapon(this.m_scene, this.m_playGameState);
                break;
            case enums.WeaponType.LASER_WEAPON:
                this.m_weapon = new CLaserWeapon(this.m_scene, this.m_playGameState);
                break;
        }

        if (this.m_weapon) {
            this.m_scene.addActor(this.m_weapon);
            this.m_weapon.activate();
            this.m_weapon.setGrade(grade);
            this.m_weapon.setOwner(this);
        }
    }

    //-------------------------------------------------------------

    public upgradeWeapon() {
        if (this.m_weapon && this.m_weapon.upgrade()) {
            return true;
        }
        return false;
    }

    //-------------------------------------------------------------

    public getWeapon() {
        return this.m_weapon;
    }

    //-------------------------------------------------------------

    public getWeaponType() {
        return this.m_weapon_type;
    }
}
export = CUpgrade;

File: Xenon-Original_C++_Code/xenon/source/upgrade.cpp
//-------------------------------------------------------------
//
// Class:	CUpgrade
//
// Author:	John M Phillips
//
// Started:	06/05/00
//
// Base:	CActor
//
// Derived:	None
//
//-------------------------------------------------------------

#include "game.h"

//-------------------------------------------------------------

CUpgrade::CUpgrade()
{
	m_offset = gsCVector(0.f,0.f);
	m_weapon = 0;
	m_weapon_type = NO_WEAPON;
}

//-------------------------------------------------------------

CUpgrade::~CUpgrade()
{
//	CActor::~CActor();
}

//-------------------------------------------------------------

bool CUpgrade::activate()
{
	if (!isActive())
		setWeapon(MISSILE_WEAPON);

	return CActor::activate();
}

//-------------------------------------------------------------

void CUpgrade::kill()
{
	if (m_weapon) {
		m_weapon->kill();
		m_weapon = 0;
		}

	CActor::kill();
}

//-------------------------------------------------------------

void CUpgrade::setOffset(const gsCVector& offset)
{
	m_offset = offset;
}

//-------------------------------------------------------------

void CUpgrade::registerHit(int energy,CActor *hitter)
{
	if (getOwner() &&
		((CShip *) getOwner())->getDiveLevel() > 0)
		return;

	CActor::registerHit(energy,hitter);
}

//-------------------------------------------------------------

void CUpgrade::onCollisionWithActor(CActor *actor)
{
	if (getOwner() &&
		((CShip *) getOwner())->getDiveLevel() > 0)
		return;

	switch (actor->getActorInfo().m_type) {
		case ACTOR_TYPE_PICKUP:
			((CPickup *) actor)->collect();
			actor->kill();
			break;
		case ACTOR_TYPE_ALIEN:
			registerHit(1,this);
			actor->registerHit(1,this);
			break;
		}
}

//-------------------------------------------------------------

void CUpgrade::onCollisionWithMap(gsCMap *map,int hits)
{
	if (getOwner() &&
		((CShip *) getOwner())->getDiveLevel() > 0)
		return;

	registerHit(UPGRADE_MAP_HIT,this);
}

//-------------------------------------------------------------

void CUpgrade::setWeapon(WeaponType type,WeaponGrade grade)
{
	if (m_weapon) {
		m_weapon->kill();
		m_weapon = 0;
		}

	m_weapon_type = type;

	switch (m_weapon_type) {
		case NO_WEAPON:
			m_weapon = 0;
			break;		
		case MISSILE_WEAPON:
			m_weapon = new CMissileWeapon;
			break;
		case HOMING_MISSILE_WEAPON:
			m_weapon = new CHomingMissileWeapon;
			break;
		case LASER_WEAPON:
			m_weapon = new CLaserWeapon;
			break;
		}

	if (m_weapon) {
		m_scene->addActor(m_weapon);
		m_weapon->activate();
		m_weapon->setGrade(grade);
		m_weapon->setOwner(this);
		}
}

//-------------------------------------------------------------

bool CUpgrade::upgradeWeapon()
{
	if (m_weapon && m_weapon->upgrade())
		return true;

	return false;
}

//-------------------------------------------------------------

CWeapon *CUpgrade::getWeapon()
{
	return m_weapon;
}

//-------------------------------------------------------------

WeaponType CUpgrade::getWeaponType()
{
	return m_weapon_type;
}

//-------------------------------------------------------------
